### PR TITLE
[ONNX] Update torch.onnx.export API

### PR DESCRIPTION
**Pull request by justinchuby**
- Add a `kwargs` option; add the `dynamic_shapes` option so users can supply it directly to `torch.export`.
- Make the options keyword-only arguments (bc-breaking)
- Deprecate the `training` and `operator_export_type` options and include a warning message. The exact time for removal is TBD but the message should discourage users from using the options.
- Deprecate two functions `exportable_ops` and pretty print export

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #131501

### CODE REVIEW

File: torch/onnx/utils.py [hunk 1]
```diff
@@ -161,10 +160,26 @@ def exporter_context(model, mode: _C_onnx.TrainingMode, verbose: bool):
...
+
+
 def export(
-    model: torch.nn.Module | torch.jit.ScriptModule | torch.jit.ScriptFunction,
+    model: torch.nn.Module
+    | torch.jit.ScriptModule
+    | torch.jit.ScriptFunction
+    | torch.export.ExportedProgram,
     args: tuple[Any, ...] | torch.Tensor,
-    f: str | io.BytesIO | None = None,
```
Comment (justinchuby) [inline torch/onnx/utils.py:167]: Still functional when input is io.BytesIO, but no longer documented.

File: test/onnx/dynamo/test_exporter_api.py [hunk 2]
```diff
@@ -263,7 +263,7 @@ def test_args_normalization_with_empty_dict_at_the_tail(self):
         )
         onnx_program_from_old_exporter = torch.onnx.export(
             SampleModelTwoInputs(),
-            (torch.randn(1, 1, 2), {"b": torch.randn(1, 1, 2)}, {}),
```
Comment (justinchuby) [inline test/onnx/dynamo/test_exporter_api.py:266]: @titaiwangms adjusted calls to make it correct

File: test/onnx/dynamo/test_exporter_api.py [hunk 3]
```diff
@@ -383,37 +383,6 @@ def test_dynamic_axes_supports_partial_dynamic_shapes(self):
             onnx_program_from_old_exporter.model_proto,
         )
```
Comment (justinchuby) [inline test/onnx/dynamo/test_exporter_api.py:385]: @titaiwangms I removed these tests as they are obsolete or too specific

File: torch/onnx/utils.py [hunk 4]
```diff
@@ -255,18 +274,17 @@ def export(
...
-            or a string containing a file name.  A binary protocol buffer will be written
-            to this file.
-        export_params (bool, default True): if True, all parameters will
+        f: Path to the output ONNX model file. E.g. "model.onnx".
+        kwargs: Named arguments to the model.
+        export_params: If True, all parameters will
             be exported. Set this to False if you want to export an untrained model.
             In this case, the exported model will first take all of its parameters
             as arguments, with the ordering as specified by ``model.state_dict().values()``
-        verbose (bool, default False): if True, prints a description of the
```
Comment (titaiwangms) [inline torch/onnx/utils.py:265]: Is default removing from doc for formatting?

File: torch/onnx/utils.py [hunk 5]
```diff
@@ -484,63 +500,84 @@ def forward(self, x):
...
-        # needs users to specify dynamic_axes
-        dynamic_shapes = _from_dynamic_axes_to_dynamic_shapes(
-            model, dynamic_axes, input_names
-        )
-        exported_program = torch.export.export(
-            model, args=args, kwargs=kwargs, dynamic_shapes=dynamic_shapes  # type: ignore[arg-type]
-        )
-        # TODO: expose ExportOptions?
+        if isinstance(model, torch.export.ExportedProgram):
+            # We the model is already exported program, so the args, kwargs, and dynamic_shapes
```
Comment (titaiwangms) [inline torch/onnx/utils.py:531]: typo

File: torch/onnx/utils.py [hunk 6]
```diff
@@ -484,63 +500,84 @@ def forward(self, x):
...
+            # are not used
+            args = ()
```
Comment (titaiwangms) [inline torch/onnx/utils.py:533]: Does this make sense? When exported program exists, we don't need args and kwargs to dynamo_export?

**Review by titaiwangms — APPROVED** [review]

File: torch/onnx/utils.py (hunk 4 above)
Comment (justinchuby) [reply torch/onnx/utils.py:265]: Yes. They are automatically added to the docs from the annotations.

File: torch/onnx/utils.py (hunk 6 above)
Comment (justinchuby) [reply torch/onnx/utils.py:533]: The args are not used so

Comment (justinchuby) [reply torch/onnx/utils.py:533]: But we can also preserve them

File: torch/onnx/utils.py (hunk 5 above)
Comment (justinchuby) [reply torch/onnx/utils.py]: Fixed

File: torch/onnx/utils.py (hunk 6 above)
Comment (titaiwangms) [reply torch/onnx/utils.py:533]: Oh I see. I guess preserving them for debugging would be better choice?

Comment (justinchuby) [reply torch/onnx/utils.py:533]: Will do

**Comment by justinchuby** [comment]
Wait for https://github.com/pytorch/pytorch/pull/131584

File: torch/onnx/utils.py (hunk 6 above)
Comment (justinchuby) [reply torch/onnx/utils.py]: Done

**Pull request closed without merging**